### PR TITLE
 Use Burmese LSTM for Burmese code point

### PR DESCRIPTION
**Pull request by makotokato**
Use Burmese LSTM model when code point is Burmese. Fixed https://github.com/unicode-org/icu4x/issues/975.

### CODE REVIEW

File: experimental/segmenter/src/line_breaker.rs [hunk 1]
```diff
@@ -245,7 +245,9 @@ fn get_break_state(left: u8, right: u8) -> i8 {
 #[inline]
 fn use_complex_breaking_utf32(codepoint: u32) -> bool {
     // Thai
-    (0xe01..=0xe7f).contains(&codepoint)
+    (0xe01..=0xe7f).contains(&codepoint) ||
+    // Burmese
+    (0x1000..=0x109f).contains(&codepoint)
```
Comment (aethanyc) [inline experimental/segmenter/src/line_breaker.rs:250]: Question: Do we need to handle these two extended range?

Myanmar Extended-A (U+AA60–U+AA7F)
Myanmar Extended-B (U+A9E0–U+A9FF)

http://www.unicode.org/faq/myanmar.html

File: experimental/segmenter/src/lstm.rs [hunk 2]
```diff
@@ -132,4 +247,37 @@ mod tests {
         let breaks = get_line_break_utf16(&text);
         assert_eq!(breaks, None, "Thai test");
     }
+
+    #[test]
+    fn burmese_word_break() {
+        const TEST_STR: &str = "မြန်မာဘာသာစကား";
+
+        let breaks = get_line_break_utf8(TEST_STR);
+        assert_eq!(breaks.unwrap(), [12, 18, 30], "Burmese test");
```
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:253]: I cannot read Burmese or Thai, so I don't have the ability to verify these break opportunity.

Question: how did you get the data? Are these break opportunities the same as ICU4C break iterator would have generated?

File: experimental/segmenter/src/lstm.rs [hunk 3]
```diff
@@ -4,21 +4,101 @@
...
+impl<'a> LanguageIterator<'a> {
+    pub fn new(input: &'a str) -> Self {
+        Self {
+            input: input.chars(),
+            last: None,
+        }
+    }
+}
+
+impl<'a> Iterator for LanguageIterator<'a> {
```
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:69]: Suggestion: Could you add a document for this iterator? It seems to iterate on substring of the same language listed in `enum Language`.

File: experimental/segmenter/src/lstm.rs [hunk 4]
```diff
@@ -86,8 +166,25 @@ impl LstmSegmenterIteratorUtf16 {
...
-    let iter = LstmSegmenterIterator::new(&*THAI_LSTM, input);
-    let result: Vec<usize> = iter.collect();
+    let mut result: Vec<usize> = Vec::new();
+    let mut lang_iter = LanguageIterator::new(&input);
+    let mut offset = 0;
+    loop {
+        let str_per_lang = lang_iter.next();
+        if str_per_lang.is_none() {
+            break;
+        }
```
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:173]: Suggestion: we can replace these lines and line 181 with `for str_per_lang in lang_iter`, right?

File: experimental/segmenter/src/lstm.rs [hunk 5]
```diff
@@ -98,8 +195,26 @@ pub fn get_line_break_utf16(input: &[u16]) -> Option<Vec<usize>> {
...
-    let iter = LstmSegmenterIteratorUtf16::new(&*THAI_LSTM, &s);
-    let result: Vec<usize> = iter.collect();
+    let mut result: Vec<usize> = Vec::new();
+    let mut lang_iter = LanguageIterator::new(&s);
+    let mut offset = 0;
+    loop {
+        let str_per_lang = lang_iter.next();
+        if str_per_lang.is_none() {
+            break;
+        }
```
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:202]: Similarly, we can use `for` loop here.

**Review by aethanyc — CHANGES_REQUESTED** [review]
Tick Request changes for my inline suggestions and questions.

File: experimental/segmenter/src/lstm.rs [hunk 6]
```diff
@@ -4,21 +4,101 @@
...
+    input: Chars<'a>,
+    last: Option<char>,
+}
+
+impl<'a> LanguageIterator<'a> {
+    pub fn new(input: &'a str) -> Self {
+        Self {
+            input: input.chars(),
+            last: None,
+        }
```
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:65]: I'm also wondering if it's possible to change `new` to the following. Maybe we could early return in the beginning of  `next()` if `last` is `None`, and simplify the code a bit.

```
        let mut input = input.chars();
        let last = input.next();
        Self { input, last }
```

File: experimental/segmenter/src/line_breaker.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (makotokato) [reply experimental/segmenter/src/line_breaker.rs:250]: Ah yes, I should add this range. And I also have to check line_break == SA.

[the remark this answers is not in the document]
Comment (sffc) [reply experimental/segmenter/src/line_breaker.rs:250]: Right; please make sure the code point ranges are consistent with the data file:

> "dic": {"\u1000": 0, "\u1001": 1, "\u1002": 2, "\u1003": 3, "\u1004": 4, "\u1005": 5, "\u1006": 6, "\u1007": 7, "\u1008": 8, "\u1009": 9, "\u100a": 10, "\u100b": 11, "\u100c": 12, "\u100d": 13, "\u100e": 14, "\u100f": 15, "\u1010": 16, "\u1011": 17, "\u1012": 18, "\u1013": 19, "\u1014": 20, "\u1015": 21, "\u1016": 22, "\u1017": 23, "\u1018": 24, "\u1019": 25, "\u101a": 26, "\u101b": 27, "\u101c": 28, "\u101d": 29, "\u101e": 30, "\u101f": 31, "\u1020": 32, "\u1021": 33, "\u1022": 34, "\u1023": 35, "\u1024": 36, "\u1025": 37, "\u1026": 38, "\u1027": 39, "\u1028": 40, "\u1029": 41, "\u102a": 42, "\u102b": 43, "\u102c": 44, "\u102d": 45, "\u102e": 46, "\u102f": 47, "\u1030": 48, "\u1031": 49, "\u1032": 50, "\u1033": 51, "\u1034": 52, "\u1035": 53, "\u1036": 54, "\u1037": 55, "\u1038": 56, "\u1039": 57, "\u103a": 58, "\u103b": 59, "\u103c": 60, "\u103d": 61, "\u103e": 62, "\u103f": 63, "\u1050": 64, "\u1051": 65, "\u1052": 66, "\u1053": 67, "\u1054": 68, "\u1055": 69, "\u1056": 70, "\u1057": 71, "\u1058": 72, "\u1059": 73, "\u105a": 74, "\u105b": 75, "\u105c": 76, "\u105d": 77, "\u105e": 78, "\u105f": 79, "\u1060": 80, "\u1061": 81, "\u1062": 82, "\u1063": 83, "\u1064": 84, "\u1065": 85, "\u1066": 86, "\u1067": 87, "\u1068": 88, "\u1069": 89, "\u106a": 90, "\u106b": 91, "\u106c": 92, "\u106d": 93, "\u106e": 94, "\u106f": 95, "\u1070": 96, "\u1071": 97, "\u1072": 98, "\u1073": 99, "\u1074": 100, "\u1075": 101, "\u1076": 102, "\u1077": 103, "\u1078": 104, "\u1079": 105, "\u107a": 106, "\u107b": 107, "\u107c": 108, "\u107d": 109, "\u107e": 110, "\u107f": 111, "\u1080": 112, "\u1081": 113, "\u1082": 114, "\u1083": 115, "\u1084": 116, "\u1085": 117, "\u1086": 118, "\u1087": 119, "\u1088": 120, "\u1089": 121, "\u108a": 122, "\u108b": 123, "\u108c": 124, "\u108d": 125, "\u108e": 126, "\u108f": 127, "\u109a": 128, "\u109b": 129, "\u109c": 130, "\u109d": 131, "\u109e": 132, "\u109f": 133, "\ua9e0": 134, "\ua9e1": 135, "\ua9e2": 136, "\ua9e3": 137, "\ua9e4": 138, "\ua9e5": 139, "\ua9e6": 140, "\ua9e7": 141, "\ua9e8": 142, "\ua9e9": 143, "\ua9ea": 144, "\ua9eb": 145, "\ua9ec": 146, "\ua9ed": 147, "\ua9ee": 148, "\ua9ef": 149, "\ua9fa": 150, "\ua9fb": 151, "\ua9fc": 152, "\ua9fd": 153, "\ua9fe": 154, "\uaa60": 155, "\uaa61": 156, "\uaa62": 157, "\uaa63": 158, "\uaa64": 159, "\uaa65": 160, "\uaa66": 161, "\uaa67": 162, "\uaa68": 163, "\uaa69": 164, "\uaa6a": 165, "\uaa6b": 166, "\uaa6c": 167, "\uaa6d": 168, "\uaa6e": 169, "\uaa6f": 170, "\uaa70": 171, "\uaa71": 172, "\uaa72": 173, "\uaa73": 174, "\uaa74": 175, "\uaa75": 176, "\uaa76": 177, "\uaa77": 178, "\uaa78": 179, "\uaa79": 180, "\uaa7a": 181, "\uaa7b": 182, "\uaa7c": 183, "\uaa7d": 184, "\uaa7e": 185, "\uaa7f": 186}

File: experimental/segmenter/src/lstm.rs [hunk 7]
```diff
@@ -4,21 +4,98 @@
...
 use std::char::decode_utf16;
+use std::str::Chars;
 
 // TODO:
 // json file is big, So I should use anoher binary format like npy.
 // But provided npy uses tensorflow dtype.
 const THAI_MODEL: &[u8; 373466] = include_bytes!(
     "../../segmenter_lstm/tests/testdata/Thai_codepoints_exclusive_model4_heavy/weights.json"
 );
+const BURMESE_MODEL: &[u8; 475209] =
```
Comment (sffc) [inline experimental/segmenter/src/lstm.rs:15]: Note: We will want this to be served by the DataProvider, which will also allow us to operate on more compact data.  This is tracked by https://github.com/unicode-org/icu4x/issues/905

File: experimental/segmenter/src/lstm.rs [hunk 8]
```diff
@@ -4,21 +4,98 @@
...
+enum Language {
+    Burmese,
+    Thai,
+    Unknown,
+}
+
+fn get_language(codepoint: u32) -> Language {
+    match codepoint {
+        0xe01..=0xe7f => Language::Thai,
+        0x1000..=0x109f => Language::Burmese,
```
Comment (sffc) [inline experimental/segmenter/src/lstm.rs:41]: Nit: Please have a single place where you define the code point ranges.

File: experimental/segmenter/src/lstm.rs (hunk 2 above)
Comment (makotokato) [reply experimental/segmenter/src/lstm.rs:253]: This is from wikipedia. This word is "Burmese Language" in Burmese. I will add comment for different of ICU4C result.

File: experimental/segmenter/src/line_breaker.rs [hunk 9]
```diff
@@ -244,10 +244,21 @@ fn get_break_state(left: u8, right: u8) -> i8 {
...
+        LineBreakRule::Strict,
+        WordBreakRule::Normal,
+    );
+    if line_break_property == SA {
+        // Thai
+        (0xe01..=0xe7f).contains(&codepoint) ||
+        // Burmese
+        (0x1000..=0x109f).contains(&codepoint) ||
+        (0xa9e0..=0xa9ff).contains(&codepoint) ||
+        (0xaa60..=0xaa7f).contains(&codepoint)
```
Comment (aethanyc) [inline experimental/segmenter/src/line_breaker.rs:258]: Nit: If Shane's nit [in previous comment ](https://github.com/unicode-org/icu4x/pull/1045#discussion_r712432252) is difficult to do, how about we leave a comment here saying something like "The code point ranges define here should be in sync with `get_language()` in lstm.rs."? Similarly, leave a comment in get_language() pointing to use_complex_breaking_utf32(), too

**Review by aethanyc — APPROVED** [review]
Thanks for simplifying the loops. Looks nicer!

File: experimental/segmenter/src/line_breaker.rs (hunk 9 above)
Comment (makotokato) [reply experimental/segmenter/src/line_breaker.rs]: I will move this to common function as `language.rs`.

**Pull request merged**